README
NAME

HOI::Match - Higher-Order Imperative "Re"features in Perl

SYNOPSIS

  use HOI::Match;

  sub point_extract {
      HOI::Match::pmatch(
          "point (x _) :: r" => sub { my %args = @_; $args{x} + point_extract($args{r}) },
          "nil" => sub { 0 }
      )->(@_)
  }

  point_extract(
      [ 
          {"type" => "point", "val" => [ 1, 2 ]},
          {"type" => "point", "val" => [ 2, 4 ]},
          {"type" => "point", "val" => [ 3, 6 ]},
      ]
  ) # we will get 6


DESCRIPTION

HOI::Match offers Erlang-like pattern matching in function parameters. 
Currently only wildcard symbols, lists and algebraic-data-type like hashrefs
are supported.

A wildcard symbol ([A-Za-z_][A-Za-z0-9_]*) matches exactly one argument.
A list is represented as an array reference. 
An algebraic-data-typed object is represented as an hashref with two keys,
namely "type", which gives its typename, and "val", which is an array reference 
containing zero or more wildcard symbols, lists, or algebraic-data-typed objects.
Multiple constructors for a given algebraic data type named A are allowed.

The BNF used to define the pattern grammar is given below:


Types: Type 
     | Type COMMA Types 


Type: CONST
    | IDENT 
    | Type CONCAT Type 
    | NIL 
    | IDENT LPAREN Typelist RPAREN 


Typelist: <eps>
        | Type Typelist 


where

CONST = (?:0(?:\.[0-9]+)?)|(?:[1-9][0-9]*(?:\.[0-9]+)?)|(?:\".+\")|(?:\'.+\')

IDENT = [A-Za-z_][A-Za-z0-9_]*

CONCAT = ::

NIL = nil

LPAREN = (

RPAREN = )

COMMA = ,

are tokens.

pmatch

The function pmatch takes an hash-formed array, which contains pattern-
subroutine pairs, where patterns are strings, sequently.

The patterns will be matched sequently. That is,

    "x, y"
    "point (_ x), y"

on arguments 
    ( { "type" => "point", "val" => [ 1, 2 ] }, 3 ) 
will be successfully matched with pattern
    "x, y" 
instead of 
    "point (_ x), y".

On a successful match, the values corresponding to identifiers in the pattern
will be passed to the subroutine in a hash. You can access them as named arguments with

    my %args = @_;

Identifiers that begin with an underscore ('_') will be ignored. They will not
be passed to the subroutine.

Version 0.07 offers aliases for identifiers in the pattern.
See the test files for details.

AUTHOR

withering <withering@cpan.org>

COPYRIGHT AND LICENSE

Copyright (C) 2014 by withering

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension HOI::Match.

0.01  Mon Feb 10 10:07:33 2014
	- original version; created by h2xs 1.23 with options
		-XAn HOI::Match
0.02  Mon Feb 10 2014
    - Fixed readme format.
0.03  Mon Feb 10 2014
    - Fixed readme format.
0.04  Wed Feb 12 2014
    - Removed unused parameters and functions. Add README and license metadata.
0.05  Wed Feb 26 2014
    - Added constant (decimal integers & floats, strings) matching.
0.06  Sat Mar 22 2014
    - Added multiple type constructor for ADTs.
0.07  Thu Dec 11 2014
    - Added alias support.
0.071 Sun Feb 15 2015
    - Removed inline code.


lib/HOI/Match.pm
package HOI::Match;

require Exporter;

use Parse::Lex;
use HOI::typeparser;

our @ISA = qw( Exporter );
our @EXPORT_OK = qw( pmatch );
our $VERSION = '0.071';

my @tokens = (
    qw (
        LPAREN      [\(]
        RPAREN      [\)]
        CONCAT      ::
        NIL         nil
        IDENT       [A-Za-z_][A-Za-z0-9_]*
        CONST       (?:0(?:\.[0-9]+)?)|(?:[1-9][0-9]*(?:\.[0-9]+)?)|(?:\".+\")|(?:\'.+\')
    ),
    COMMA => q/,/
);

my $lexer = Parse::Lex->new(@tokens);
$lexer->skip('\s+');
my $parser = HOI::typeparser->new();

sub lexana {
    my $token = $lexer->next;
    if (not $lexer->eoi) {
        return ($token->name, $token->text);
    } else {
        return ('', undef);
    }
}

my %compiled_patterns;

sub pcompile {
    $lexer->from(shift);
    $parser->YYParse(yylex => \&lexana)
}

sub astmatch {
    my ($ast, $args) = @_;
    return (0, {}) if ($#{$ast} ne $#{$args});
    my %switches = (
        "const" =>
        sub {
            my ($sym, $val) = @_;
            if( (substr($sym, 0, 1) eq '\'') or (substr($sym, 0, 1) eq '"') ) {
                my $quote = substr($sym, 0, 1);
                return ($sym eq $quote.$val.$quote) ? (1, {}) : (0, {});
            } else {
                return ($sym == $val) ? (1, {}) : (0, {});
            }
        },
        "any" => 
        sub { 
            my ($sym, $val) = @_; 
            (1, ((substr($sym, 0, 1) ne '_') ? { $sym => $val } : {})) 
        },
        "list" => 
        sub { 
            my ($l, $val) = @_; 
            if (($#{$l} >= 0) and ($#{$val} >= 0)) {
                my ($s1, $r1) = astmatch([ $l->[0] ], [ $val->[0] ]);
                my ($s2, $r2) = astmatch([ $l->[1] ], [ [ @$val[1..$#{$val}] ] ]);
                return ($s1 * $s2, { %$r1, %$r2 });
            } elsif (($#{$l} < 0) and ($#{$val} < 0)) {
                return (1, {});
            } else {
                return (0, {});
            }
        },
        "adt" =>
        sub { 
            my ($adt, $val) = @_;
            return (0, {}) if ((not defined $val->{"type"}) or (not defined $val->{"val"}));
            my ($sym, $typelist) = ($adt->[0], $adt->[1]);
            return (0, {}) if ($adt->[0] ne $val->{"type"});
            return (0, {}) if ($#{$adt->[1]} != $#{$val->{"val"}});
            astmatch($adt->[1], $val->{"val"})
        }
    );
    my $ret = {};
    for (my $idx = 0; $idx <= $#{$ast}; $idx++) {
        my ($status, $result) = $switches{$ast->[$idx]->{"kind"}}->($ast->[$idx]->{"val"}, $args->[$idx]);
        if ($status) {
            $ret = { %$ret, %$result };
        } else {
            return (0, {})
        }
    }
    (1, $ret)
}

sub pmatch {
    my $patterns = \@_;
    sub {
        my $args = \@_;
        while (@$patterns) {
            my $pattern = shift @$patterns;
            my $handler = shift @$patterns;
            my $pattern_sig = (caller(1))[3].$pattern;
            $compiled_patterns{$pattern_sig} = pcompile($pattern) if (not defined $compiled_patterns{$pattern_sig});
            my $pattern_ast = $compiled_patterns{$pattern_sig};
            my ($status, $results) = astmatch($pattern_ast, $args);
            if ($status) {
                my ($package) = caller(1);
                local $AttrPrefix = $package.'::';
                #attr $results;
                my $evalstr = '';
                for my $key (keys %$results) {
                    $evalstr .= 'local $'."$AttrPrefix"."$key".' = $results->{'."$key".'}; ';
                }
                return eval "{ $evalstr ".'$handler->(%$results); }';
            }
        }
        0
    }
}

1;
__END__

=head1 NAME

HOI::Match - Higher-Order Imperative "Re"features in Perl

=head1 SYNOPSIS

  use HOI::Match;

  sub point_extract {
      HOI::Match::pmatch(
          "point (x _) :: r" => sub { my %args = @_; $args{x} + point_extract($args{r}) },
          "nil" => sub { 0 }
      )->(@_)
  }

  point_extract(
      [ 
          {"type" => "point", "val" => [ 1, 2 ]},
          {"type" => "point", "val" => [ 2, 4 ]},
          {"type" => "point", "val" => [ 3, 6 ]},
      ]
  ) # we will get 6


=head1 DESCRIPTION

HOI::Match offers Erlang-like pattern matching in function parameters. 
Currently only wildcard symbols, lists and algebraic-data-type like hashrefs
are supported.

A wildcard symbol ([A-Za-z_][A-Za-z0-9_]*) matches exactly one argument.
A list is represented as an array reference. 
An algebraic-data-typed object is represented as an hashref with two keys,
namely "type", which gives its typename, and "val", which is an array reference 
containing zero or more wildcard symbols, lists, or algebraic-data-typed objects.
Multiple constructors for a given algebraic data type named A are allowed.

The BNF used to define the pattern grammar is given below:


Types: Type 
     | Type COMMA Types 


Type: CONST
    | IDENT 
    | Type CONCAT Type 
    | NIL 
    | IDENT LPAREN Typelist RPAREN 


Typelist: <eps>
        | Type Typelist 


where

CONST = (?:0(?:\.[0-9]+)?)|(?:[1-9][0-9]*(?:\.[0-9]+)?)|(?:\".+\")|(?:\'.+\')

IDENT = [A-Za-z_][A-Za-z0-9_]*

CONCAT = ::

NIL = nil

LPAREN = (

RPAREN = )

COMMA = ,

are tokens.

=head2 pmatch

The function pmatch takes an hash-formed array, which contains pattern-
subroutine pairs, where patterns are strings, sequently.

The patterns will be matched sequently. That is,

    "x, y"
    "point (_ x), y"

on arguments 
    ( { "type" => "point", "val" => [ 1, 2 ] }, 3 ) 
will be successfully matched with pattern
    "x, y" 
instead of 
    "point (_ x), y".

On a successful match, the values corresponding to identifiers in the pattern
will be passed to the subroutine in a hash. You can access them as named arguments with

    my %args = @_;

Identifiers that begin with an underscore ('_') will be ignored. They will not
be passed to the subroutine.

Version 0.07 offers aliases for identifiers in the pattern.
See the test files for details.

=head1 AUTHOR

withering <withering@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by withering

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/HOI/typeparser.pm
####################################################################
#
#    This file was generated using Parse::Yapp version 1.05.
#
#        Don't edit this file, use source file instead.
#
#             ANY CHANGE MADE HERE WILL BE LOST !
#
####################################################################
package HOI::typeparser;
use vars qw ( @ISA );
use strict;

@ISA= qw ( Parse::Yapp::Driver );
use Parse::Yapp::Driver;



sub new {
        my($class)=shift;
        ref($class)
    and $class=ref($class);

    my($self)=$class->SUPER::new( yyversion => '1.05',
                                  yystates =>
[
	{#State 0
		ACTIONS => {
			'NIL' => 3,
			'CONST' => 5,
			'IDENT' => 2
		},
		GOTOS => {
			'Types' => 1,
			'Type' => 4
		}
	},
	{#State 1
		ACTIONS => {
			'' => 6
		}
	},
	{#State 2
		ACTIONS => {
			'LPAREN' => 7
		},
		DEFAULT => -4
	},
	{#State 3
		DEFAULT => -6
	},
	{#State 4
		ACTIONS => {
			'CONCAT' => 8,
			'COMMA' => 9
		},
		DEFAULT => -1
	},
	{#State 5
		DEFAULT => -3
	},
	{#State 6
		DEFAULT => 0
	},
	{#State 7
		ACTIONS => {
			'IDENT' => 2,
			'NIL' => 3,
			'CONST' => 5
		},
		DEFAULT => -8,
		GOTOS => {
			'Typelist' => 10,
			'Type' => 11
		}
	},
	{#State 8
		ACTIONS => {
			'NIL' => 3,
			'CONST' => 5,
			'IDENT' => 2
		},
		GOTOS => {
			'Type' => 12
		}
	},
	{#State 9
		ACTIONS => {
			'IDENT' => 2,
			'NIL' => 3,
			'CONST' => 5
		},
		GOTOS => {
			'Type' => 4,
			'Types' => 13
		}
	},
	{#State 10
		ACTIONS => {
			'RPAREN' => 14
		}
	},
	{#State 11
		ACTIONS => {
			'NIL' => 3,
			'CONST' => 5,
			'IDENT' => 2,
			'CONCAT' => 8
		},
		DEFAULT => -8,
		GOTOS => {
			'Type' => 11,
			'Typelist' => 15
		}
	},
	{#State 12
		ACTIONS => {
			'CONCAT' => 8
		},
		DEFAULT => -5
	},
	{#State 13
		DEFAULT => -2
	},
	{#State 14
		DEFAULT => -7
	},
	{#State 15
		DEFAULT => -9
	}
],
                                  yyrules  =>
[
	[#Rule 0
		 '$start', 2, undef
	],
	[#Rule 1
		 'Types', 1,
sub
#line 11 "typeparser.yp"
{ [ $_[1] ] }
	],
	[#Rule 2
		 'Types', 3,
sub
#line 12 "typeparser.yp"
{ my $sublist = $_[3]; my @list= ($_[1], @$sublist); \@list }
	],
	[#Rule 3
		 'Type', 1,
sub
#line 15 "typeparser.yp"
{ { "kind" => "const", "val" => $_[1] } }
	],
	[#Rule 4
		 'Type', 1,
sub
#line 16 "typeparser.yp"
{ { "kind" => "any", "val" => $_[1] } }
	],
	[#Rule 5
		 'Type', 3,
sub
#line 17 "typeparser.yp"
{ { "kind" => "list", "val" => [ $_[1], $_[3] ] } }
	],
	[#Rule 6
		 'Type', 1,
sub
#line 18 "typeparser.yp"
{ { "kind" => "list", "val" => [] } }
	],
	[#Rule 7
		 'Type', 4,
sub
#line 19 "typeparser.yp"
{ { "kind" => "adt", "val" => [ $_[1], $_[3] ] } }
	],
	[#Rule 8
		 'Typelist', 0,
sub
#line 22 "typeparser.yp"
{ [] }
	],
	[#Rule 9
		 'Typelist', 2,
sub
#line 23 "typeparser.yp"
{ my $sublist = $_[2]; my @list= ($_[1], @$sublist); \@list }
	]
],
                                  @_);
    bless($self,$class);
}

#line 26 "typeparser.yp"


1;
